src/metrics/metrics.go

```diff
@@ -272,6 +220,64 @@ func populateCustomMetrics(instanceEntity *integration.Entity, connection *conne
 	}
 }
 
+// metricsFromCustomQueryRow obtains a map of metrics from a row resulting from a custom query.
+// A particular metric can be configured either with:
+// - Specific columns in the query: metric_name, metric_type, metric_value
+// - The corresponding `query.Name` and `query.Type`
+// When both are defined, the query columns have precedence. Besides, if type is not defined it is automatically deteced.
+// The rest of the query columns are also taken as metrics/attributes (detecting their types automatically).
+// Besides, if `query.Prefix` is defined, all metric and attribute names will include the corresponding prefix.
+func metricsFromCustomQueryRow(row []string, columns []string, query customQuery) (map[string]customQueryMetricValue, error) {
+	metrics := map[string]customQueryMetricValue{}
+
+	var metricValue string
+	metricType := query.Type
+	metricName := query.Name
+
+	for i, columnName := range columns {
+		switch columnName {
+		case "metric_name":
+			metricName = row[i]
+		case "metric_type":
+			metricType = row[i]
+		case "metric_value":
+			metricValue = row[i]
+		default:
+			name := query.Prefix + columnName
+			value := row[i]
+			metrics[name] = customQueryMetricValue{value: value, sourceType: detectMetricType(value)}
+		}
+	}
+
+	if metricValue == "" {
+		if metricName != "" {
+			return nil, fmt.Errorf("%w: name %q, query %q", errMissingMetricNameCustomQuery, metricName, query.Query)
+		}
+		return metrics, nil
+	}
+
+	if metricName == "" {
+		return nil, fmt.Errorf("%w: query %q", errMissingMetricValueCustomQuery, query.Query)
+	}
+
+	var sourceType metric.SourceType
+	if metricType != "" {
+		sourceTypeFromQuery, err := metric.SourceTypeForName(metricType)
+		if err != nil {
+			return nil, fmt.Errorf("invalid metric type %s: %w", metricType, err)
+		}
+		sourceType = sourceTypeFromQuery
+	} else {
+		sourceType = detectMetricType(metricValue)
+	}
+
+	metricName = query.Prefix + metricName
+
+	metrics[metricName] = customQueryMetricValue{value: metricValue, sourceType: sourceType}
```

Comment:
why returning a map if there is always just one element?

Reply:
Can we do che following:
 - have a function just to scan the row looking for metric_name and metric_type
 - have a second function to scan the row whenever those are missing
Then merging the two sets of metrics.
Even though is faster it is a bit confusing that while we look for the metric_name we save other metrics

Reply:
You are right, there were to many things going on in the same function. Instead of the proposed refactor I updated it as follows:

- The function extracting info from the row/columns and return the map of metrics is kept
- The logic to build the metric from the 'special columns' is extracted to a different function.

This way, we avoid going through the whole array twice and the function is a bit easier to read. Let me know if it helps.